File: anagram_finder.cpp
#include <unordered_map>
#include <iostream>
#include <fstream>

using namespace std;

unordered_map<char, int> word_to_map(string word) {
    unordered_map<char, int> word_map;

    for (char ch: word) {
        if (word_map.find(ch) == word_map.end()) {
            word_map[ch] = 1;
        } else {
            word_map[ch]++;
        };
    };

    return word_map;
};

void lower(string &s) {
    for (int i = 0; i < s.length(); i++) {
        s[i] = (char)tolower(s[i]);
    };
};

bool is_anagram(unordered_map<char, int> mp, string word) {
    unordered_map<char, int> target_map = word_to_map(word);
    if (mp == target_map) {
        return true;
    } else {
        return false;
    };
};


int main() {
    unordered_map<char, int> word_map;
    string word;
    string fname;

    cout << "Enter word: ";
    cin >> word;

    cout << "Enter filename: ";
    cin >> fname;

    lower(word);
    lower(fname);
    word_map = word_to_map(word);

    ifstream file(fname);
    string line;

    while (getline(file, line)) {
        lower(line);
        if (is_anagram(word_map, line)) {
            cout << line << endl;
        };
    };

    file.close();
    return 0;
}

File: anagram_finder.py
def word_to_dict(word):
    word_dict = dict()
    for ch in word:
        if ch not in word_dict:
            word_dict[ch] = 1
        else:
            word_dict[ch] += 1
    return word_dict


def is_anagram(word1, word2):
    return word_to_dict(word1) == word_to_dict(word2)


def main():
    print("Enter word: ", end='')
    target = input()
    print("Enter file name: ", end='')
    fname = input()
    with open(fname, 'r') as file:
        for line in file:
            if is_anagram(line.strip().lower(), target):
                print(line.strip())


if __name__ == '__main__':
    main()

